examples/directml/llm/llm.py

```diff
@@ -23,6 +23,110 @@
+    "lm_head.linear": "lm_head",
+    "final_layernorm": "norm",
+    # "layers": "model.layers",
+    "transformer": "model",
+    ".h.": ".layers.",
+    "ln": "input_layernorm",
+    "mixer": "self_attn",
+    "Wqkv": "query_key_value",
+    "dense": "o_proj",
+}
+
+def map_key(origin_key):
+    for k, v in PHI_MAPPING.items():
+        if k in origin_key:
+            origin_key = origin_key.replace(k, v)
+    return origin_key
+
+def convert_weights(original_weights, mapping, config):
+    converted_weights = {}
+    original_weights_keys = sorted(original_weights.keys())
+
+    for original_weights_key in original_weights_keys:
+        new_key = original_weights_key
+
+        if "rotary_emb" in new_key:
+            continue
+
+        if "Wqkv" in new_key:
+            if "weight" in new_key:
+                weight = original_weights[new_key]
+                weights_shape = weight.shape
+                weight = (
+                    weight.view(3, config.num_heads, -1, config.hidden_size)
+                    .transpose(0, 1)
+                    .reshape(*weights_shape)
+                ).view(config.num_heads, 3, -1)
+                q_proj_key = map_key(new_key.replace("Wqkv", "q_proj"))
+                k_proj_key = map_key(new_key.replace("Wqkv", "k_proj"))
+                v_proj_key = map_key(new_key.replace("Wqkv", "v_proj"))
+                converted_weights[q_proj_key] = weight[:, 0, ...].reshape(2560, 2560)
+                converted_weights[k_proj_key] = weight[:, 1, ...].reshape(2560, 2560)
+                converted_weights[v_proj_key] = weight[:, 2, ...].reshape(2560, 2560)
+
+                original_weights.pop(new_key)
+
+            elif "bias" in new_key:
+                bias = original_weights[new_key]
+                bias_shape = bias.shape
+                bias = bias.view(3, config.num_heads, -1).transpose(0, 1).reshape(*bias_shape).view(config.num_heads, 3, -1)
+                q_proj_key = map_key(new_key.replace("Wqkv", "q_proj"))
+                k_proj_key = map_key(new_key.replace("Wqkv", "k_proj"))
+                v_proj_key = map_key(new_key.replace("Wqkv", "v_proj"))
+                converted_weights[q_proj_key] = bias[..., 0, :].reshape(2560)
+                converted_weights[k_proj_key] = bias[..., 1, :].reshape(2560)
+                converted_weights[v_proj_key] = bias[..., 2, :].reshape(2560)
+
+                original_weights.pop(new_key)
+
+            continue
+
+        new_key = map_key(new_key)
+
+        converted_weights[new_key] = original_weights.pop(original_weights_key)
+
+    return converted_weights
+
+def _download(url: str, root: str):
+    repo_id = f"{url.split('/')[3]}/{url.split('/')[4]}"
+    filename = f"{url.split('/')[-1]}"
+    hf_hub_download(
+        repo_id=repo_id,
+        filename=filename,
+        force_filename=root,
+        local_dir_use_symlinks=False,
+    )
+
+def load_phi2_checkpoint(checkpoint_path):
```

Comment:
Why is this necessary? We should be able to automatically download the weights the same way we do for other LLMs, without having to interact and manually load the safetensors files.